pymongo/ssl_support.py

```diff
@@ -49,12 +52,24 @@
     import ssl as _stdlibssl  # noqa: F401
     from ssl import CERT_NONE, CERT_REQUIRED
 
-    HAS_SNI = _ssl.HAS_SNI
     IPADDR_SAFE = True
+
+    if HAVE_PYSSL:
+        HAS_SNI = _pyssl.HAS_SNI | _ssl.HAS_SNI
```

Comment:
This logic for HAS_SNI is problematic since this happens at import time but the actual ssl impl at runtime (stdlib vs pyopenssl) may or may not support SNI. I believe we need something like:
```python
        # We have to pass hostname / ip address to wrap_socket
        # to use SSLContext.check_hostname.
        if ssl_context.has_sni:
            ...
```

Reply:
Hmm I don't think I understand the problem? HAS_SNI was previously defined in the same spot. Did my changes introduce this problem? Or did the previous code also have the same problem?

Reply:
The new logic is incorrect. If pyopenssl is install and `_pyssl.HAS_SNI=False` and `_ssl.HAS_SNI=True`, the sync code will incorrectly think that pyopenssl supports SNI.

Reply:
Ah, okay i understand how the logic is incorrect now (I previously thought that the two would be the same value regardless of pyopenssl vs stdlibssl) 
would you mind elaborating on your suggested fix though? I'm not sure I understand how that would resolve pyopenssl's has_sni and stdlib ssl's has_sni being two different values and how to navigate having those two potentially different values in the rest of the code?

Reply:
Not sure if I did it the way you envisioned it, but I created a helper function, `has_sni(is_sync) -> bool`, and call that everywhere `HAS_SNI` was being used.